docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,421 @@
+increase user choice. Therefore we should remain agnostic at the protocol level
+about exactly what such metadata looks like, instead we should build a single
+API for applications to use metadata associated with a project. The schemas and
+interpretations of these data types then become composable at the application
+layer.
+
+## Design Goals
+
+- Local first
+- Extensible. It should be possible to extend Radicle with new collaborative
+  data types without changing the protocol version
+- Interoperable. There should be a straightforward API for disparate tools to
+  interact with collaborative objects
+
+Local first because you should not need to be online to modify collaborative
+objects and no one should be able to stop you from changing data that is on
+your computer. That said, collaborative objects are _collaborative_, users need
+to be able to receive changes to the same data from other people and this
+raises the problem of how to merge those changes. I see two options here:
+
+- CRDTs, data structures which allow conflict free merges
+- Application level merges. Much as Git requires user action to merge
+  conflicting changes, we could expose the merge points of collaborative
+  objects to the user to resolve.
+
+The latter here is undesirable for a lot of reasons but most relevant is that
+it conflicts with extensibility and interoperability. in order to provide a good 
+UX tools would need to provide a UI for users to do a three way merge of the 
+data they are working on. This contradicts the interoperability goal.
+
+This leaves us with CRDTs. For the purposes of this RFC I am going to assume
+that we will use the [Automerge](https://github.com/automerge/automerge) CRDT
+implementation. Automerge is capable of merging arbitrary JSON objects (the
+data model is actually richer than JSON, with types for byte arrays, various
+precision integers and floats etc.). An alternative approach would be to write
+a custom CRDT for each data type we want to replicate, see [##Alternative
+Approaches] for a discussion of this design.
+
+## Example: Issues
+
+To motivate and contextualise this proposal we introduce a running example
+which is dear to many peoples hearts; issues. We imagine this from the
+perspective of the application developer. For expositional purposes we assume
+that the application communicates with a radicle implementation via an HTTP
+RPC. However, the HTTP RPC is not a proposal of this RFC, the proposed API will
+be specified in terms of `librad` later in this document.
+
+The first thing the developer must do is decide on the schema of their data and
+represent it as a JSON schema. We use this simple schema:
+
+```json
+{
+  "type": "object"
+  "properties": {
+    "title": {
+        "type": "string"
+    },
+    "description": {
+        "type": "string"
+    },
+    "author": {
+        "type": "string",
+        "description": "The radicle ID of the author of the issue"
+    },
+    "signature": {
```

Comment:
In this simple schema you could sign over a canonical JSON encoding of the "title", "description", "author", and "id". To make this more realistic we could add an ID field to each comment and the sign over ("comment ID", "comment text", "parent_id"), where "parent_id" is the ID of the previous comment according to the author at the time they wrote it. Validating such a signature would involve asserting that the previous comment appears before the given comment in the document.

If maintaining a stricter invariant than precedence is necessary then you could encode the comments as a tree structure rather than a list.

There are obviously a whole bunch of other problems - specifying what the signature algorithm is, what key was used etc. Hopefully these can all be concerns of the particular data type implementation and the collaborative objects infrastructure is agnostic about them. One thing I have thought about that might be useful would be to add a custom schema type which verifies that a signature is a valid signature with respect to another URN in the schema. Maybe something like this:

```json
{
    "signature": {
        "type": "rad_signature",
        "signing_urn_path": "/author",
        "over": [
            "id",
            "title",
            "description",
            "author",
        ]
        "description": "A base64 encoded signature of the issue"
    },
}
```

I feel like in the process of trying to implement the issues feature we will encounter problems which will help flesh out what such features should look like.

Reply:
> In this simple schema you could sign over a canonical JSON encoding of the "title", "description", "author", and "id". To make this more realistic we could add an ID field to each comment and the sign over ("comment ID", "comment text", "parent_id"), where "parent_id" is the ID of the previous comment according to the author at the time they wrote it. Validating such a signature would involve asserting that the previous comment appears before the given comment in the document.

To me, this seems like responsibilities are being mixed. I understand that having such a signature would be useful but I would imagine it being placed somewhere _above_ the payload, e.g. a git trailer. And so the signature is always over the payload itself -- where a reply to a particular comment could be in the schema itself.

I'm also wondering if that would be covered by the `rad/signed_refs` (after my comment blunder :))?

Reply:
Yes, I also feel like this would put too much burden on the application layer. I read the proposal as we never actually store the JSON representation, only opaque automerge changes, so it's not so easy to see how to attribute a particular change to an author -- but otoh, an author can change everything about the object, including authorship fields.

I think this can probably be solved if we're not using `tree`s, but the actual git history. So I'll wait for that and inquire again later =]

> covered by the rad/signed_refs

Nope, that has a different purpose. It attests the _origin_ of a set of branches, not authorship.